homeassistant/components/dhcp/__init__.py

```diff
@@ -134,14 +114,135 @@ def process_updated_address_data(self, ip_address, data):
 
             _LOGGER.debug("Matched %s against %s", data, entry)
 
-            self.hass.add_job(
+            self.add_job(
                 self.hass.config_entries.flow.async_init(
                     entry["domain"],
                     context={"source": DOMAIN},
                     data={IP_ADDRESS: ip_address, **data},
                 )
             )
 
+    @abstractmethod
+    def add_job(self, job):
+        """Pass a job to based on which context we are in."""
+
+
+class DeviceTrackerWatcher(WatcherBase):
+    """Class to watch dhcp data from routers."""
+
+    def __init__(self, hass, address_data, integration_matchers):
+        """Initialize class."""
+        super().__init__(hass, address_data, integration_matchers)
+        self._unsub = None
+
+    @callback
+    def async_stop(self):
+        """Stop watching for new device trackers."""
+        if self._unsub:
+            self._unsub()
+            self._unsub = None
+
+    @callback
+    def async_start(self):
+        """Stop watching for new device trackers."""
+        self._unsub = async_track_state_added_domain(
+            self.hass, [DEVICE_TRACKER_DOMAIN], self._async_process_device_event
+        )
+        for state in self.hass.states.async_all(DEVICE_TRACKER_DOMAIN):
+            self._async_process_device_state(state)
+
+    @callback
+    def _async_process_device_event(self, event: Event):
+        """Process a device tracker state change event."""
+        self._async_process_device_state(event.data.get("new_state"))
+
+    @callback
+    def _async_process_device_state(self, state: State):
+        """Process a device tracker state."""
+        if state.state != STATE_HOME:
+            return
+
+        attributes = state.attributes
+
+        if attributes.get(ATTR_SOURCE_TYPE) != SOURCE_TYPE_ROUTER:
+            return
+
+        ip_address = attributes.get(ATTR_IP)
+        hostname = attributes.get(ATTR_HOST_NAME)
+        mac_address = attributes.get(ATTR_MAC)
+
+        if ip_address is None or hostname is None or mac_address is None:
+            return
+
+        self.process_client(ip_address, hostname, _format_mac(mac_address))
+
+    def add_job(self, job):
+        """Pass a job to async_add_job since we are in async context."""
+        self.hass.async_add_job(job)
```

Comment:
Currently the only job we have is a task. Not sure if there are other plans for thr future.
```suggestion
        self.hass.async_create_task(job)
```